src/lcm/process_model.py

```diff
@@ -288,14 +336,99 @@ def _get_function_with_dummy_params(func):
+    def next_func(*args, **kwargs):  # noqa: ARG001
+        return grid
+
+    return next_func
+
+
+def _get_stochastic_weight_function(raw_func, name, variable_info, grids):
+    """Get a function that returns the transition weights of a stochastic variable.
+
+    Example:
+    Consider a stochastic variable 'health' that takes two values {0, 1}. The transition
+    matrix is thus 2x2. We create the weighting function and then select the weights
+    that correspond to the case where 'health' is 0.
+
+    >>> from lcm.mark import StochasticInfo
+    >>> def next_health(health):
+    >>>     pass
+    >>> next_health._stochastic_info = StochasticInfo()
+    >>> params = {"shocks": {"health": np.arange(4).reshape(2, 2)}}
+    >>> weight_func = _get_stochastic_weight_function(
+    >>>     raw_func=next_health,
+    >>>     name="health"
+    >>>     variable_info=variable_info,
+    >>>     grids=grids,
+    >>> )
+    >>> weight_func(health=0, params=params)
+    >>> array([0, 1])
+
+
+    Args:
+        raw_func (callable): The raw next function of the stochastic variable.
+        name (str): The name of the stochastic variable.
+        variable_info (pd.DataFrame): A table with information about model variables.
+        grids (dict): Dictionary containing all variables of the model. The keys are
+            the names of the variables. The values are the grids.
+
+    Returns:
+        callable: A function that returns the transition weights of the stochastic
+            variable.
+
+    """
+    function_parameters = list(inspect.signature(raw_func).parameters)
+
+    # Assert that stochastic next function only depends on state variables
+    for arg in function_parameters:
+        if not variable_info.loc[arg, "is_state"]:
+            raise ValueError(
+                f"Stochastic variables can only depend on state variables, but {name} "
+                f"depends on {arg}.",
+            )
+
+    label_translators = _get_label_translators(
+        variables=function_parameters,
+        grids=grids,
+    )
+
+    new_kwargs = [*function_parameters, "params"]
+
+    @with_signature(args=new_kwargs)
+    def weight_func(*args, **kwargs):
+        kwargs = all_as_kwargs(args, kwargs, arg_names=new_kwargs)
+        indices = [
+            label_translators[arg](**{arg: kwargs[arg]}) for arg in function_parameters
+        ]
+        return kwargs["params"]["shocks"][name][*indices]
+
+    return weight_func
+
+
+def _get_label_translators(variables, grids):
+    """Get a dictionary of label translators.
+
+    Args:
+        variables (list): List of variable names.
+        grids (dict): Dictionary containing all variables of the model. The keys are
+            the names of the variables. The values are the grids.
+
+    Returns:
+        dict: Dictionary that maps variable names to label translators.
+
+    """
+    return {
+        var: _get_label_translator(labels=grids[var], in_name=var) for var in variables
+    }
```

Comment:
@janosg Can you check whether this is the intended usage of the label translator defined in `function_evaluator.py`?

Reply:
In principle yes, but I think in `weight_func` I would use `all_as_args` instead of `all_as_kwargs` and then simplify the rest a bit. 

Also: took me quite a while to see the plural-s on `_get_label_translators`. We need a more visible distinction. 

And probably make `_get_label_translator` public then?